### PR TITLE
Issue 338 - Unit Tests Run Slow

### PR DESCRIPTION
**Pull request by lprichar**
# Summary

Introduces the option of writing light-weight, fast-running unit tests in addition to the current in-memory database style integration tests.  The key to this functionality is the new `.AsRepository()` extension method (see below).

# Details

* Moves all existing test infrastructure into an `IntegrationTests\` folder and adjusts namespaces
* Adds a `\UnitTests` folder for placing fast tests into
* Adds a `AbpProjectNameUnitTestBase` and `AbpProjectNameUnitTestModule` into `\UnitTests` that unit tests will inherit from and use
* Renames `AbpProjectNameTestBase` to `AbpProjectNameIntegrationTestBase` to keep the base classes clearly differentiated
* Adds classes: TestQueryable, TestAsyncQueryProvider, TestAsyncEnumerable, and TestAsyncEnumerator, all of which provide the infrastructure for LINQ queries against in-memory lists of data
* Adds an `AsRepository()` extension method that creates working IRepository mocks given lists of objects (or single objects)
* Adds several unit tests for UserAppService into `UnitTests\Users\UserAppService_Tests.cs` primarily for documentation purposes

# To Test

1. Run the tests in `UnitTests\Users\UserAppService_Tests.cs`

Expected: They should all pass

### CODE REVIEW

File: aspnet-core/.gitignore [hunk 1]
```diff
@@ -250,3 +250,9 @@ paket-files/
 # JetBrains Rider
 .idea/
 *.sln.iml
+
+# NCrunch
```
Comment (lprichar) [inline aspnet-core/.gitignore]: Oh, and I added NCrunch to the .gitignore.  It's not technically related to the PR, and I'm happy to separate it out if you like.

File: aspnet-core/test/AbpCompanyName.AbpProjectName.Tests/UnitTests/FakeServices/FakeLogger.cs [hunk 2]
```diff
@@ -0,0 +1,36 @@
+﻿using System;
+using System.Collections.Generic;
+using Castle.Core.Logging;
+
+namespace AbpCompanyName.AbpProjectName.Tests.UnitTests.FakeServices
+{
+    public class FakeLogger : NullLogger, ILogger
```
Comment (lprichar) [inline aspnet-core/test/AbpCompanyName.AbpProjectName.Tests/UnitTests/FakeServices/FakeLogger.cs]: Technically I guess the FakeLogger isn't really required for this PR, but I used it a lot for verifying logging in my unit tests.  I'm happy to remove it if you like.

File: aspnet-core/test/AbpCompanyName.AbpProjectName.Tests/UnitTests/Users/UserAppService_Tests.cs [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        private UserAppService GetUserAppService(Role role, User user = null)
+        {
+            var roleRepository = role.AsRepository<Role>();
+            var userRepository = user.AsRepository<User, long>();
+            return GetUserAppService(roleRepository, userRepository);
+        }
+
+        private UserAppService GetUserAppService(IRepository<Role> roleRepository, IRepository<User, long> userRepository)
+        {
+            // todo: extract an interface for UserManager and use it in UserAppService to simplify mocking and remove all the null args
```
Comment (lprichar) [inline aspnet-core/test/AbpCompanyName.AbpProjectName.Tests/UnitTests/Users/UserAppService_Tests.cs]: I don't know if adding an interface for UserManager and RoleManager etc is on the roadmap, but using their concrete classes in the constructor injection makes pure unit testing much harder and more brittle.

**Comment by ismcagdas** [comment]
Thanks a lot @lprichar 👍 
We will check it for v4.2 milestone.

**Comment by ismcagdas** [comment]
@lprichar we have reviewed your PR and it is very good I think but we thought that this will be a bit complex for a startup template. We will keep this PR open and use it as an example for the ones who face the same performance problem on their proejcts.

Thanks again for your great contribution 😄 .

**Pull request closed without merging**